man/unicoverage.1
.TH UNICOVERAGE 1 "2007 Dec 31"
.SH NAME
unicoverage \- Print coverage of each Unicode Script
.SH SYNOPSIS
\fBunicoverage \fP[\-i\fIinput-file\fP] [\-o\fIoutput-file\fP] [\-n]
.SH DESCRIPTION
.B unicoverage
reads a GNU Unifont .hex font and uses data in
.B coverage.dat
(which must reside in the current directory).  The output is
the percent coverage of each script listed in the coverage.dat
file.  The Unicode code points in the input .hex file must be
in ascending order.
.SH OPTIONS
.TP 12
\fB\-i\fP
Specify the input file. The default is STDIN.
.TP
\fB\-n\fP
Print the number of glyphs in each range instead of
a percentage count.
.TP
\fB\-o\fP
Specify the output file. The default is STDOUT.
.SH EXAMPLE
Sample usage:
.PP
.RS
unicoverage < unifont.hex >coverage.txt
.RE
.SH FILES
coverage.dat, *.hex GNU Unifont files.
.SH SEE ALSO
.BR bdfimplode (1),
.BR hex2bdf (1),
.BR hex2otf (1),
.BR hex2sfd (1),
.BR hexbraille (1),
.BR hexdraw (1),
.BR hexkinya (1),
.BR hexmerge (1),
.BR johab2syllables (1),
.BR johab2ucs2 (1),
.BR unibdf2hex (1),
.BR unibmp2hex (1),
.BR unibmpbump (1),
.BR unidup (1),
.BR unifont (5),
.BR unifont-johab631 (5),
.BR unifont-viewer (1),
.BR unifont1per (1),
.BR unifontchojung (1),
.BR unifontksx (1),
.BR unifontpic (1),
.BR unigen-hangul (1),
.BR unigencircles (1),
.BR unigenwidth (1),
.BR unihex2bmp (1),
.BR unihex2png (1),
.BR unihexfill (1),
.BR unihexgen (1),
.BR unihexpose (1),
.BR unihexrotate (1),
.BR unijohab2html (1),
.BR unipagecount (1),
.BR unipng2hex (1)
.SH AUTHOR
.B unicoverage
was written by Paul Hardy.
.SH LICENSE
.B unicoverage
is Copyright \(co 2007, 2008 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known real bugs exist, except that this software does not perform
extensive error checking on its input files.


NEWS
2024-02-24: Release 15.1.05<F2>
  * Added almost 600 CJK ideographs, including:
    - The entire CJK Unified Ideographs Extension D range of 222 glyphs
    - 335 commonly used Cantonese ideographs in Plane 2 and Plane 3.
    - Several updates to CJK Unified Ideographs Extension I.
  * All Hangul glyphs now have a consistent look with the Hangul Syllables,
    with new glyyphs for:
    - U+1100..U+11FF Hangul Jamo
    - U+3130..U+318F Hangul Compatibility Jamo
    - U+A960..U+A97F Hangul Jamo Extended-A
    - U+D7B0..U+D7FF Hangul Jamo Extended-B.
  * Several other glyph modifications; see the ChangeLog file for details.

2023-10-29: Release 15.1.04
  * Added larger Plane 2 supersets to default and Japanese font files.

2023-10-21: Release 15.1.03
  * Boris Zhang and Yzy32767 contributed CJK Unified Ideographs
    Extension I glyphs (U+2EBF0..U+2EE5D).
  * 湖 远星 contributed 14 glyph updates to CJK Unified Ideographs
    Extension B and Extension C: U+2A6D7..U+2A6DF and
    U+2B735..U+2B739.

2023-09-21: Release 15.1.02
   * 湖 远星 adjusted 46 glyphs in the Wen Quan Yi range, U+2F00..U+9FFF, and
     contributed CJK Unified Ideographs Extension G glyphs U+30000..U+3017F.
   * Added new man pages to MAN1PAGES & MAN5PAGES definitions for installation.

2023-09-12: Release 15.1.01
   * Glyphs updated for Unicode 15.1.0 release.
   * OpenType fonts are now built by default, replacing TrueType fonts.
     TrueType fonts can be built manually by invoking "make truetype"
     in the font directory.
   * New Hangul Johab 6/3/1 encoding devised by Ho-Seok Ee for building
     the Hangul Syllables block (U+AC00..U+D7A3), with new supporting
     programs johab2syllables, unigen-hangul, and unijohab2html.  This
     encoding replaces Unifont's original Hanterm-derived encoding.
     New source files hangul.h and unihangul-support.c provide support
     for rendering syllable bitmaps with all Unicode Hangul letters,
     including all ancient Hangul, through the new unigen-hangul program.
     These functions also can support user-written Hangul syllable rendering
     programs, for example on embedded systems using graphics displays.
   * Added three hexadecimal digit representation CSUR/UCSUR scripts:
     - U+EBE0..U+EBEF: Boby Lapointe's "bibi-binary" notation
     - U+EBF0..U+EBFF: Bruce Alan Martin's bit location notation
     - U+ECF0..U+ECFF: Ronald O. Whitaker's triangular notation.
   * Modified hexdraw to support bitmap manipulation of up to 64 rows
     by 128 columns for 64-by-128 graphics display screen drawing support.
     Added the unihexpose program to transpose Unifont glyphs, for use
     with graphics display controller chips that encode each byte as
     8 rows by 1 column, left to right.
   * Other minor updates; see ChangeLog for details.

2023-06-04: Release 15.0.06
   * Further adjustments to Hangul glyphs in the Basic Multilingual Plane.

2023-06-03: Release 15.0.05
   * New Hangul Syllables block (U+AC00..U+D7A3) contributed
     by Minseo Lee.  Generated with his modifications of three
     combining patterns in the johab2ucs2 Perl script.
   * Modified U+5829 and U+6FF9 Chinese glyphs.
   * Full details in ChangeLog file.

2023-05-28: Release 15.0.04
   * Adjustments to Greek and Coptic letters.
   * Adjustments to two heart glyphs.
   * Added Engsvanáli (U+E100..U+E14F) CSUR glyphs.

2023-05-21: Release 15.0.03
   * Fixed a typo in src/Makefile that could affect
     "make install".
   * Install unifont_all.hex in $(PKGDEST) directory.
   * Other minor changes; see ChangeLog file for details.

2023-05-20: Release 15.0.02
   * N.B.: This is the last planned version to build TrueType
     font files as part of the standard font build with

           make BUILDFONT=1

     In the future, only OpenType font files will be part of
     the ordinary font building using the hex2otf program.
     TrueType fonts will have to be built manually using

           cd font
           make truetype

   * Minor update to allow for separate building of TrueType
     and OpenType fonts.  This is in preparation for the future
     removal of TrueType font files from default font builds,
     in preference for OpenType font files.  That change is
     planned for the release of Unifont 15.1.01, which will
     correspond to the release of the Unicode 15.1.0 standard
     scheduled for September 2023.
   * Added Doxygen annotation to C programs and corresponding
     ".h" header files.  Annotated output is generated in a
     new "doxygen" subdirectory.  Doxygen output consists of
     a browsable set of HTML pages and a hyperlinked PDF document.
   * Updates to make files to fix issues with redundant building
     under some circumstances.
   * Added 196 Plane 2 and Plane 3 Chinese ideographs drawn by
     湖 远星 to the basic Unifont font files for coverage of
     the General Standard Chinese Characters set.  Also added
     biang glyphs drawn by Ming Fan.
   * Fullwidth CJK characters that appeared to be boldface in
     the range U+FF00..U+FFEF were updated using glyphs from
     Wen Quan Yi, with some modifications.

2022-09-13: Release 15.0.01
   * Update for Unicode 15.0.0 release.
   * "make install" now installs Unifont OpenType font files in the
     same directory as TrueType fonts by default.  The intention is
     to remove TrueType fonts file from the "install" target for the
     Unicode 16.0.0 release unless there is a reason to keep them.
   * Existing Unicode scripts in Unifont maintain complete coverage.
   * Under ConScript Unicode Registry (UCSUR) scripts added:
     - U+E1D0..U+E1FF Syai
     - U+E360..U+E37F Sarkai
     - U+ED80..U+EDAF Iranic.
   * New Unicode 15.0.0 Plane 1 scripts:
     - U+10EFD..U+10EFF Arabic Extended-C
     - U+11B00..U+11B5F Devanagari Extended-A
     - U+11F00..U+11F5F Kawi
     - U+1D2C0..U+1D2DF Kaktovik Numerals
     - U+1E030..U+1E08F Cyrillic Exntended-D
     - U+1E4D0..U+1E4FF Nag Mundari.

2022-06-04: Release 14.0.04

src/unicoverage.c
/**
   @file unicoverage.c

   @brief unicoverage - Show the coverage of Unicode plane scripts
                        for a GNU Unifont hex glyph file

   @author Paul Hardy, unifoundry <at> unifoundry.com, 6 January 2008
   
   @copyright Copyright (C) 2008, 2013 Paul Hardy

   Synopsis: unicoverage [-ifont_file.hex] [-ocoverage_file.txt]

   This program requires the file "coverage.dat" to be present
   in the directory from which it is run.
*/
/*
   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
   2016 (Paul Hardy): Modified in Unifont 9.0.01 release to remove non-existent
   "-p" option and empty example from help printout.

   2018 (Paul Hardy): Modified to cover entire Unicode range, not just Plane 0.

   11 May 2019: [Paul Hardy] changed strcpy function call to strlcpy
   for better error handling.

   31 May 2019: [Paul Hardy] replaced strlcpy call with strncpy
   for compilation on more systems.

   4 June 2022: [Paul Hardy] Adjusted column spacing for better alignment
   of Unicode Plane 1-15 scripts.  Added "-n" option to print number of
   glyphs in each range instead of percent coverage.

   18 September 2022: [Paul Hardy] in nextrange function, initialize retval.

   21 October 2023: [Paul Hardy]
   Added full function prototype for nextrange function in main function.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#define MAXBUF 256   ///< Maximum input line length - 1


/**
   @brief The main function.

   @param[in] argc The count of command line arguments.
   @param[in] argv Pointer to array of command line arguments.
   @return This program exits with status 0.
*/
int
main (int argc, char *argv[])
{

   int      print_n=0;        /* print # of glyphs, not percentage   */
   unsigned i;                /* loop variable                       */
   unsigned slen;             /* string length of coverage file line */
   char     inbuf[256];       /* input buffer                        */
   unsigned thischar;         /* the current character               */

   char *infile="", *outfile="";  /* names of input and output files        */
   FILE *infp, *outfp;        /* file pointers of input and output files    */
   FILE *coveragefp;          /* file pointer to coverage.dat file          */
   int cstart, cend;          /* current coverage start and end code points */
   char coverstring[MAXBUF];  /* description of current coverage range      */
   int nglyphs;               /* number of glyphs in this section           */

   /* to get next range & name of Unicode glyphs */
   int nextrange (FILE *coveragefp, int *cstart, int *cend, char *coverstring);

   void print_subtotal (FILE *outfp, int print_n, int nglyphs,
                        int cstart, int cend, char *coverstring);

   if ((coveragefp = fopen ("coverage.dat", "r")) == NULL) {
      fprintf (stderr, "\nError: data file \"coverage.dat\" not found.\n\n");
      exit (0);
   }

   if (argc > 1) {
      for (i = 1; i < argc; i++) {
         if (argv[i][0] == '-') {  /* this is an option argument */
            switch (argv[i][1]) {
               case 'i':  /* name of input file */
                  infile = &argv[i][2];
                  break;
               case 'n':  /* print number of glyphs instead of percentage */
                  print_n = 1;
               case 'o':  /* name of output file */
                  outfile = &argv[i][2];
                  break;
               default:   /* if unrecognized option, print list and exit */
                  fprintf (stderr, "\nSyntax:\n\n");
                  fprintf (stderr, "   %s -p<Unicode_Page> ", argv[0]);
                  fprintf (stderr, "-i<Input_File> -o<Output_File> -w\n\n");
                  exit (1);
            }
         }
      }
   }
   /*
      Make sure we can open any I/O files that were specified before
      doing anything else.
   */
   if (strlen (infile) > 0) {
      if ((infp = fopen (infile, "r")) == NULL) {
         fprintf (stderr, "Error: can't open %s for input.\n", infile);
         exit (1);
      }
   }
   else {
      infp = stdin;
   }
   if (strlen (outfile) > 0) {
      if ((outfp = fopen (outfile, "w")) == NULL) {
         fprintf (stderr, "Error: can't open %s for output.\n", outfile);
         exit (1);
      }
   }
   else {
      outfp = stdout;
   }

   /*
      Print header row.
   */
   if (print_n) {
      fprintf (outfp, "# Glyphs      Range        Script\n");
      fprintf (outfp, "--------      -----        ------\n");
   }
   else {
      fprintf (outfp, "Covered      Range        Script\n");
      fprintf (outfp, "-------      -----        ------\n\n");
   }

   slen = nextrange (coveragefp, &cstart, &cend, coverstring);
   nglyphs = 0;

   /*
      Read in the glyphs in the file
   */
   while (slen != 0 && fgets (inbuf, MAXBUF-1, infp) != NULL) {
      sscanf (inbuf, "%x", &thischar);

      /* Read a character beyond end of current script. */
      while (cend < thischar && slen != 0) {
         print_subtotal (outfp, print_n, nglyphs, cstart, cend, coverstring);

         /* start new range total */
         slen = nextrange (coveragefp, &cstart, &cend, coverstring);
         nglyphs = 0;
      }
      nglyphs++;
   }

   print_subtotal (outfp, print_n, nglyphs, cstart, cend, coverstring);

   exit (0);
}

/**
   @brief Get next Unicode range.

   This function reads the next Unicode script range to count its
   glyph coverage.

   @param[in] coveragefp File pointer to Unicode script range data file.
   @param[in] cstart Starting code point in current Unicode script range.
   @param[in] cend Ending code point in current Unicode script range.
   @param[out] coverstring String containing <cstart>-<cend> substring.
   @return Length of the last string read, or 0 for end of file.
*/
int
nextrange (FILE *coveragefp,
              int *cstart, int *cend,
              char *coverstring)
{
   int i;
   static char inbuf[MAXBUF];
   int retval;         /* the return value */

   retval = 0;

   do {
      if (fgets (inbuf, MAXBUF-1, coveragefp) != NULL) {
         retval = strlen (inbuf);
         if ((inbuf[0] >= '0' && inbuf[0] <= '9') ||
             (inbuf[0] >= 'A' && inbuf[0] <= 'F') ||
             (inbuf[0] >= 'a' && inbuf[0] <= 'f')) {
            sscanf (inbuf, "%x-%x", cstart, cend);
            i = 0;
            while (inbuf[i] != ' ') i++;  /* find first blank */
            while (inbuf[i] == ' ') i++;  /* find next non-blank */
            strncpy (coverstring, &inbuf[i], MAXBUF);
         }
         else retval = 0;
      }
      else retval = 0;
   } while (retval == 0 && !feof (coveragefp));

   return (retval);
}


/**
   @brief Print the subtotal for one Unicode script range.

   @param[in] outfp Pointer to output file.
   @param[in] print_n 1 = print number of glyphs, 0 = print percentage.
   @param[in] nglyphs Number of glyphs in current range.
   @param[in] cstart Starting code point for current range.
   @param[in] cend Ending code point for current range.
   @param[in] coverstring Character string of "<cstart>-<cend>".
*/
void print_subtotal (FILE *outfp, int print_n, int nglyphs,
                     int cstart, int cend, char *coverstring) {

   /* print old range total */
   if (print_n) {  /* Print number of glyphs, not percentage */
      fprintf (outfp, " %6d ", nglyphs);
   }
   else {
      fprintf (outfp, " %5.1f%%", 100.0*nglyphs/(1+cend-cstart));
   }

   if (cend < 0x10000)
      fprintf (outfp, "  U+%04X..U+%04X   %s",
               cstart, cend, coverstring);
   else
      fprintf (outfp, " U+%05X..U+%05X  %s",
               cstart, cend, coverstring);

   return;
}
